[Lua Yi Da] iP

luayida99 opened the pull request:
## Side 
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Side frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
-  fun
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. forget about it and let it manage your tasks for you 😉

And it is most importantly **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

luayida99 committed:
Add greet, echo, exit functionality

luayida99 committed:
Add and list functionality added

luayida99 committed:
Mark as done functionality implemented

luayida99 committed:
Add todo, event, deadline

luayida99 committed:
Added automated text UI testing

luayida99 committed:
Gave Side some personality, checked for todo empty description

luayida99 committed:
update UI test to reflect new changes

luayida99 committed:
add custom exceptions, level 5 completed

luayida99 committed:
add delete functionality, refactor some code to abstract out functionalities

luayida99 committed:
fix bug where passing noninteger to done and delete crashes program

luayida99 committed:
update ui test to reflect new functions

luayida99 committed:
Change some hardcoded errors to use exceptions, handled case where no index specified for done and delete

luayida99 committed:
Add javadocs

luayida99 committed:
use enum for commands

luayida99 committed:
standardise code style for switch statements

luayida99 committed:
Add save to hard disk functionality

Storage class to model saving to hard disk using
text file written to everytime bye is called.

luayida99 committed:
Merge pull request #1 from luayida99/branch-Level-7

Add save to hard disk functionality

luayida99 committed:
Add datetime

luayida99 committed:
Merge pull request #2 from luayida99/branch-Level-8

Add datetime

luayida99 committed:
Abstract similar code into classes

luayida99 committed:
Package code

luayida99 committed:
Add JUnit tests

luayida99 committed:
Add jar file

luayida99 committed:
Add javadocs

luayida99 committed:
Follow java coding standards

luayida99 committed:
Merge pull request #4 from luayida99/branch-A-CodingStandard

Follow java coding standards

luayida99 committed:
Add back deleted file to resolve conflict

luayida99 committed:
Merge pull request #3 from luayida99/branch-A-JavaDoc

Add javadocs

luayida99 committed:
Add find functionality

luayida99 committed:
Merge pull request #5 from luayida99/branch-Level-9

Add find functionality

luayida99 committed:
Remove unused exception class

luayida99 committed:
Remove unused exception class

luayida99 committed:
Merge remote-tracking branch 'origin/master'

luayida99 committed:
Add javadocs

luayida99 committed:
Merge pull request #6 from luayida99/branch-Level-9

Remove unused exception class

luayida99 committed:
Edit javadocs

luayida99 committed:
Refactor code to reduce unnecessary function calls

luayida99 committed:
Edit javadocs

luayida99 committed:
Add javadocs

luayida99 committed:
Merge pull request #7 from luayida99/add-gradle-support

Add Gradle support

luayida99 committed:
Integrate gradle for running Side and tests

luayida99 committed:
Merge pull request #8 from luayida99/branch-A-Gradle

Integrate gradle for running Side and tests

luayida99 committed:
Implement checkstyle plugin

luayida99 committed:
Merge pull request #9 from luayida99/branch-A-CheckStyle

Implement checkstyle plugin

luayida99 committed:
Implement basic GUI for Side

luayida99 committed:
Merge pull request #10 from luayida99/branch-Level-10

Implement basic GUI for Side

luayida99 committed:
Fix dialog box cutoff issue

Dialog box used to be cut off with too many tasks.
This commit fixes this issue.

luayida99 committed:
Add javadocs for side.gui classes

luayida99 committed:
Update javadocs

e0543403 commented on src/main/java/side/util/Ui.java:
```diff
@@ -0,0 +1,210 @@
+            if (datetimeArr.length < 2) {
+                throw new WrongFormatException("event [task name] /at [YYYY-MM-DD], [HHMM] /to [YYYY-MM-DD], [HHMM]\n"
+                        + "event [task name] /at [YYYY-MM-DD] /to [YYYY-MM-DD]");
+            }
```
Perhaps more line breaks to improve readability?

luayida99 replied:
thanks for the suggestion, will look into it!

e0543403 commented on src/main/java/side/util/Ui.java:
```diff
@@ -0,0 +1,210 @@
+            TooManyIndexesException {
+        if (input.split("\\s+").length == 2) {
+            int taskNum = this.parser.tryIntParsing(input.split("\\s+")[1]);
+            if (taskNum > taskList.length() || taskNum <= 0) {
```
Perhaps abstract this conditional outside the if statement?

e0543403 commented on src/main/java/side/tasks/Event.java:
```diff
@@ -0,0 +1,102 @@
+        String[] startDatetimeArr = startInput.split("\\s+");
+        String[] endDatetimeArr = endInput.split("\\s+");
+
+        if (startDatetimeArr.length == 2 && endDatetimeArr.length == 2) {
```
Maybe abstract out this condition and reasoning?

e0543403 commented on src/main/java/side/util/Ui.java:
```diff
@@ -0,0 +1,210 @@
+        if (input.contains("/by") && (this.parser.findDeadlineDatetime(input) != null)) {
+            String datetime = this.parser.findDeadlineDatetime(input);
+            String description = this.parser.findDescription(input);
+            taskList.addDeadline(description, datetime);
```
Perhaps add a line break?

e0543403 reviewed (commented):
I could not really find any code quality violations albeit some parts were quite dense or complex. Overall it looks really cool and unique, a job well done.

YaleChen299 commented on src/main/java/side/Side.java:
```diff
@@ -0,0 +1,88 @@
+
+    /**
+     * Launches GUI.
+     * @param args Standard main method args.
```
Maybe an empty line after the description before the parameter section will improve readability and align with the standard.

YaleChen299 commented on src/main/java/side/tasks/Deadline.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    @Override
+    public String toString() {
+        StringBuilder deadlineLine = new StringBuilder();
```
This naming `deadlineLine` may be confusing to read at first, maybe you can name it like `deadlineString` or `deadlineMessage`.

YaleChen299 commented on src/main/java/side/tasks/Event.java:
```diff
@@ -0,0 +1,102 @@
+     * @return End datetime for instance.
+     */
+    public DateTime getEndDatetime() {
+        return this.startDatetime;
```
Maybe this is a small mis? According to the method name, we are expecting the `endDatetime`.

luayida99 replied:
good catch, thanks!

YaleChen299 commented on src/main/java/side/util/Storage.java:
```diff
@@ -0,0 +1,119 @@
+                taskLine.append(taskDetails);
+            } else {
+                taskLine.append("T | No time");
+            }
```
Maybe you can apply polymorphism in this part as well, just like how you did for the printed output.

YaleChen299 commented on src/main/java/side/util/TaskList.java:
```diff
@@ -0,0 +1,188 @@
+     */
+    public void save() {
+        try {
+            String filepath = "./data";
```
This path is seen in many places, maybe it would be a good idea to keep it as a higher level constant so as to prevent error in spelling and make it easier to change the path in the future?

YaleChen299 reviewed (commented):
Thanks for the hard work! Generally very good work with a lot of detailed java doc description for the classes and methods. Left a few comments on some minor nits.

luayida99 committed:
Fix bug with endDateTime

luayida99 committed:
Edit javadocs

hmanhduc2k commented on src/main/java/side/gui/GuiWindow.java:
```diff
@@ -0,0 +1,70 @@
+    private final Image userImage = new Image(this.getClass().getResourceAsStream("/images/user.jpg"));
+    private final Image sideImage = new Image(this.getClass().getResourceAsStream("/images/side.png"));
+
+    @FXML
+    public void initialize() {
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
+    }
+
+    @FXML
+    private void handleUserInput() {
+        Side side = new Side();
+        String input = userInput.getText();
+        String response = side.replyToCommand(userInput.getText(), taskList);
+        dialogContainer.getChildren().addAll(DialogBox.getUserDialog(input, userImage),
+                DialogBox.getSideDialog(response, sideImage));
+
+        //pauses program for 1 second upon bye
+        if (userInput.getText().stripTrailing().equals("bye")) {
+            PauseTransition delay = new PauseTransition(Duration.seconds(1));
+            delay.setOnFinished(event -> Platform.exit());
+            delay.play();
+        }
+
+        userInput.clear();
+    }
```
Would be great to consider adding documentation (JavaDoc) to all these methods to explain what they do and what kind of changes they handle to the user input.

luayida99 replied:
ah, i did javadocs for public methods but this could be something to look into, thanks

hmanhduc2k commented on src/main/java/side/util/Ui.java:
```diff
@@ -206,8 +184,9 @@ private String convertFindList(ArrayList<Task> list) {
+    public String handleFind(String input, TaskList taskList) {
         if (input.split("\\s+").length == 1) {
             throw new NoIndexException();
         } else {
```
Your JavaDoc comment should consider adding a @throw statement to explain why you are throwing the NoIndexException() in this code block (or what the exception is about in the first place)

hmanhduc2k commented on src/main/resources/view/GuiWindow.fxml:
```diff
@@ -0,0 +1,20 @@
+<?import javafx.scene.layout.AnchorPane?>
+<?import javafx.scene.layout.AnchorPane?>
+<?import javafx.scene.layout.VBox?>
+
+<AnchorPane maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity" prefHeight="600.0" prefWidth="400.0" xmlns="http://javafx.com/javafx/8.0.171" xmlns:fx="http://javafx.com/fxml/1" fx:controller="side.gui.GuiWindow">
+    <children>
+        <TextField fx:id="userInput" layoutY="558.0" onAction="#handleUserInput" prefHeight="41.0" prefWidth="324.0" AnchorPane.bottomAnchor="1.0" />
+        <Button fx:id="sendButton" layoutX="324.0" layoutY="558.0" mnemonicParsing="false" onAction="#handleUserInput" prefHeight="41.0" prefWidth="76.0" text="Send" />
+        <ScrollPane fx:id="scrollPane" hbarPolicy="NEVER" hvalue="1.0" prefHeight="557.0" prefWidth="400.0" vvalue="1.0">
+            <content>
+                <VBox fx:id="dialogContainer" prefHeight="552.0" prefWidth="388.0" />
+            </content>
+        </ScrollPane>
+    </children>
+</AnchorPane>
```
Just a personal suggestion - try to break down the FXML field declaration to multiple lines (so that no line exceed 100 characters) and hence making the code more readable.

hmanhduc2k commented on src/main/java/side/gui/DialogBox.java:
```diff
@@ -0,0 +1,68 @@
+import javafx.scene.layout.HBox;
+
+public class DialogBox extends HBox {
+
+    @FXML
+    private Label dialog;
+    @FXML
+    private ImageView displayPicture;
+
```
Illegal pattern trailing whitespace

hmanhduc2k reviewed (commented):
Overall - your code seems good, no major standard violation is found, documentation is done in a careful manner. It was quite hard to find any mistake, but it would be perfect if you can improve the readability of your code by a bit, specifically your FXML file, and adding more documentations. Great jobs :)

luayida99 committed:
Use assert to document key assumptions

There are some assumptions in the codebase that
are currently undocumented.

Documenting these assumptions will make it
easier for us to check for violations which
could break some of the existing functionality.

Let's use the assert feature to document these
assumptions.

luayida99 committed:
Improve code quality

There are some minor issues with code quality
pointed out by reviewers.

Fixing these issues would improve overall code
quality.

Let's:
* fix issues pointed out by reviewers
* Examine code and find any other issues, and fix
those issues as well

luayida99 committed:
Merge pull request #11 from luayida99/branch-A-Assertions

Use assert to document key assumptions

luayida99 committed:
Merge branch 'master' into branch-A-CodeQuality

luayida99 committed:
Merge pull request #12 from luayida99/branch-A-CodeQuality

Improve code quality

luayida99 committed:
Rename UI method for clarity

luayida99 committed:
Implement better search

Searching for a perfect match in find function
is unrealistic, especially when there are many
tasks.

Modifying the search functionality to search
for substrings in description makes find more
versatile

Let's implement a search that checks for any
matching substring in all existing task descriptions

luayida99 committed:
Merge pull request #15 from luayida99/branch-BCD-Extension

Implement better search

luayida99 committed:
Call toString for datetime conversion

luayida99 committed:
Improve code quality

Some conditions are still not abstracted out.

Abstracting these conditions will improve code readablity.

Let's extract these conditions and improve
overall code quality.

luayida99 committed:
Improve GUI

luayida99 committed:
Setup project website

luayida99 committed:
Set theme jekyll-theme-slate

luayida99 committed:
Set theme jekyll-theme-cayman

luayida99 committed:
Improve code quality based on cs2103-bot comment

luayida99 committed:
Merge remote-tracking branch 'origin/master'

luayida99 committed:
Update user guide
